cmd/allocator/v1alpha1/allocation.proto

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package v1alpha1;
+
+import "k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto";
+import "google/api/annotations.proto";
+
+service AllocationService {
+ rpc PostAllocate(AllocationRequest) returns (AllocationResponse) {
+   option (google.api.http) = {
+     post: "/v1alpha1/gameserverallocation"
+     body: "*"
+   };
+ }
+}
+
+message AllocationRequest {
+  // The k8s namespace that is hosting the targetted fleet of gameservers to be allocated
+  string namespace = 1;
+
+  // If specified, multi-cluster policies are applied. Otherwise, allocation will happen locally.
+  MultiClusterSetting multiClusterSetting = 2;
+
+  // The required allocation. Defaults to all GameServers.
+  k8s.io.apimachinery.pkg.apis.meta.v1.LabelSelector requiredGameServerSelector = 3;
+
+  // The ordered list of preferred allocations out of the `required` set.
+  // If the first selector is not matched, the selection attempts the second selector, and so on.
+  repeated k8s.io.apimachinery.pkg.apis.meta.v1.LabelSelector preferredGameServerSelector = 4;
+
+  // Scheduling strategy. Defaults to "Packed".
+  SchedulingStrategy scheduling = 5;
+  enum SchedulingStrategy {
+    Packed = 0;
+    Distributed = 1;
+  }
+
+  // MetaPatch is optional custom metadata that is added to the game server at
+  // allocation You can use this to tell the server necessary session data
+  MetaPatch metaPatch = 6;
+}
+
+message AllocationResponse {
+  GameServerAllocationState state = 1;
+
+  // The allocation state
+  // TODO: instead of using state return 500 for unallocated and 409 for contention
```

Comment:
We should talk about this TODO more - you can see some background here, about why we made this decision:
https://github.com/googleforgames/agones/issues/485
https://github.com/googleforgames/agones/issues/436

We very deliberately didn't have an unfound game server be an error. I'm not sure it's appropriate here?

Also, I think it will be confusing to have 2 different approaches for inside cluster vs outside at this stage.

Reply:
Thanks for the explanation. For a RESTful API, the http status represents the status of the result and to me receiving 200 for an unsuccessful response is contradicting. In kubernetes world, status is a good representation of the state of a CRD, but this does not apply to http response standards. 

I think the allocation REST API should return a meaningful http status in case of failure instead of 200. For now, I removed the TODO and opened an issue to discuss this further.

Reply:
That's makes sense :+1: 

Will add some notes to the issue.